src/util/configInheritance.tsx

```diff
@@ -0,0 +1,246 @@
+import { InstanceAndProfileFormValues } from "components/forms/instanceAndProfileFormValues";
+import { LxdProfile } from "types/profile";
+import { CreateInstanceFormValues } from "pages/instances/CreateInstance";
+import { EditInstanceFormValues } from "pages/instances/EditInstance";
+import { isDiskDevice, isNicDevice } from "util/devices";
+import { LxdDiskDevice, LxdNicDevice } from "types/device";
+import { ProjectFormValues } from "pages/projects/CreateProject";
+import { ConfigurationRowFormikValues } from "components/ConfigurationRow";
+import { ConfigField } from "types/config";
+import { useQuery } from "@tanstack/react-query";
+import { queryKeys } from "util/queryKeys";
+import { fetchConfigOptions } from "api/server";
+import { toConfigFields } from "util/config";
+import { getInstanceKey } from "util/instanceConfigFields";
+import { useParams } from "react-router-dom";
+import { fetchProfiles } from "api/profiles";
+import { getProjectKey } from "util/projectConfigFields";
+import { StorageVolumeFormValues } from "pages/storage/forms/StorageVolumeForm";
+import { fetchStoragePool } from "api/storage-pools";
+import { getStorageVolumeDefault, getVolumeKey } from "util/storageVolume";
+import { getNetworkDefault } from "util/networks";
+
+export interface ConfigRowMetadata {
+  value?: string;
+  source: string;
+  configField?: ConfigField;
+}
+
+export const getConfigRowMetadata = (
+  values: ConfigurationRowFormikValues,
+  name: string,
+): ConfigRowMetadata => {
+  switch (values.type) {
+    case "instance":
+    case "profile":
+      return getInstanceRowMetadata(values, name);
+    case "project":
+      return getProjectRowMetadata(values, name);
+    case "storageVolume":
+      return getStorageVolumeRowMetadata(values, name);
+    case "network":
+      return getNetworkRowMetadata(name);
+  }
+};
+
+const getInstanceRowMetadata = (
+  values: InstanceAndProfileFormValues | ProjectFormValues,
+  name: string,
+): ConfigRowMetadata => {
+  const { data: configOptions } = useQuery({
```

Comment:
Although the queries are cached, still feels a bit weird that this request code is here. Wouldn't it make more sense to move the logic higher up before individual config rows are constructed and pass a configOptions lookup down to the child?

Reply:
Yes, I agree. Though currently, only instances, profiles and projects use the dynamic config lookup. Networks and storages don't need them. But both are planned to be added to it soon in the near future of this cycle. As soon as both are changed towards it, I am happy to move this query up the chain, as we'll need it in all cases anyway.

So for now let's keep the lookup here, in the future, let's keep in mind to move it up to the ConfigRow component.